File: Core/graph_rag.py
"""
Core/graph_rag.py — Graph RAG with Community Detection & Summarization.

Industry best practice: Beyond basic entity-triple storage, Graph RAG
builds community-level summaries enabling answers to global/thematic
queries that no single chunk can address.

Pipeline:
  1. Extract entities and relationships from chunks
  2. Build graph communities (connected components / Leiden)
  3. Generate summaries per community
  4. At query time: match communities → use summaries as context
  5. For local queries: traditional entity lookup + multi-hop

This is inspired by Microsoft's "GraphRAG" (2024) paper which showed
that community summaries dramatically improve answers to global
sensemaking questions ("What are the main themes in this dataset?").

References:
  - Edge et al. "From Local to Global: A Graph RAG Approach" (Microsoft, 2024)
  - Community detection via connected components (lightweight alternative to Leiden)
"""

from __future__ import annotations

import hashlib
import logging
import re
from collections import defaultdict
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional, Set, Tuple

logger = logging.getLogger(__name__)


@dataclass
class Community:
    """A community of related entities from the knowledge graph."""

    community_id: str
    entities: List[str] = field(default_factory=list)
    relationships: List[Tuple[str, str, str]] = field(default_factory=list)
    summary: str = ""
    keywords: List[str] = field(default_factory=list)
    size: int = 0
    level: int = 0  # Hierarchy level (0 = leaf, higher = more abstract)


@dataclass
class GraphRAGResult:
    """Result from Graph RAG query."""

    answer: str = ""
    matched_communities: List[Community] = field(default_factory=list)
    entity_context: List[Dict[str, Any]] = field(default_factory=list)
    strategy: str = "local"  # "local" or "global"
    community_summaries_used: int = 0


class GraphRAG:
    """Graph-based RAG with community detection and hierarchical summarization.

    Extends the basic KnowledgeGraph with:
    - Automatic community detection (connected components)
    - LLM-generated community summaries
    - Global query answering via community map-reduce
    - Local query answering via entity + multi-hop traversal

    Usage:
        graph_rag = GraphRAG(knowledge_graph=kg, llm_fn=my_generate)
        graph_rag.build_communities()  # After indexing documents

        # Global query (themes, overview)
        result = graph_rag.query("What are the main themes?", strategy="global")

        # Local query (specific entity)
        result = graph_rag.query("How does X relate to Y?", strategy="local")
    """

    def __init__(
        self,
        knowledge_graph: Any = None,
        llm_fn: Optional[Callable] = None,
        min_community_size: int = 3,
        max_community_summary_len: int = 500,
    ):
        """
        Args:
            knowledge_graph: KnowledgeGraph instance for entity/triple data
            llm_fn: function(prompt) -> str for summarization
            min_community_size: minimum entities to form a community
            max_community_summary_len: max chars per community summary
        """
        self.kg = knowledge_graph
        self.llm_fn = llm_fn
        self.min_community_size = min_community_size
        self.max_summary_len = max_community_summary_len
        self.communities: List[Community] = []
        self._entity_to_community: Dict[str, str] = {}

    def build_communities(self) -> List[Community]:
        """Detect communities via connected components and generate summaries.

        Steps:
          1. Extract adjacency from knowledge graph
          2. Find connected components (lightweight community detection)
          3. Filter by min size
          4. Generate summaries per community
        """
        if not self.kg:
            logger.warning("[GraphRAG] No knowledge graph provided")
            return []

        # Step 1: Build adjacency list from KG
        adjacency, entity_names = self._build_adjacency()
        if not adjacency:
            logger.info("[GraphRAG] Empty graph, no communities to build")
            return []

        # Step 2: Find connected components
        components = self._connected_components(adjacency, entity_names)

        # Step 3: Filter and build Community objects
        self.communities = []
        self._entity_to_community = {}

        for i, (entities, relationships) in enumerate(components):
            if len(entities) < self.min_community_size:
                continue

            cid = hashlib.sha256(
                "|".join(sorted(entities)).encode()
            ).hexdigest()[:12]

            community = Community(
                community_id=cid,
                entities=sorted(entities),
                relationships=relationships,
                size=len(entities),
                keywords=self._extract_keywords(entities, relationships),
            )

            # Generate summary
            if self.llm_fn:
                community.summary = self._generate_community_summary(community)
            else:
                community.summary = self._heuristic_summary(community)

            self.communities.append(community)
            for entity in entities:
                self._entity_to_community[entity.lower()] = cid

        logger.info(
            f"[GraphRAG] Built {len(self.communities)} communities "
            f"from {sum(c.size for c in self.communities)} entities"
        )
        return self.communities

    def query(
        self,
        query: str,
        strategy: str = "auto",
        top_k_communities: int = 3,
    ) -> GraphRAGResult:
        """Answer a query using the knowledge graph and community summaries.

        Args:
            query: User query
            strategy: "local" (entity lookup), "global" (community summaries),
                      or "auto" (detect best strategy)
            top_k_communities: max communities for global queries
        """
        if strategy == "auto":
            strategy = self._detect_strategy(query)

        if strategy == "global":
            return self._global_query(query, top_k_communities)
        else:
            return self._local_query(query)

    def _global_query(self, query: str, top_k: int) -> GraphRAGResult:
        """Answer using community summaries (map-reduce pattern).

        1. Match query to relevant communities
        2. Collect community summaries
        3. Use summaries as context for generation
        """
        matched = self._match_communities(query, top_k)

        if not matched:
            return GraphRAGResult(strategy="global")

        # Build context from community summaries
        summaries = []
        for community in matched:
            if community.summary:
                summaries.append(
                    f"[Community: {', '.join(community.entities[:5])}]\n"
                    f"{community.summary}"
                )

        context = "\n\n".join(summaries)
        answer = ""

        if self.llm_fn and context:
            try:
                prompt = (
                    f"Using the following knowledge graph community summaries, "
                    f"answer the question comprehensively.\n\n"
                    f"Community Summaries:\n{context}\n\n"
                    f"Question: {query}\n\nAnswer:"
                )
                answer = self.llm_fn(prompt)
            except Exception as e:
                logger.warning(f"[GraphRAG] Global generation failed: {e}")

        return GraphRAGResult(
            answer=answer or "",
            matched_communities=matched,
            strategy="global",
            community_summaries_used=len(matched),
        )

    def _local_query(self, query: str) -> GraphRAGResult:
        """Answer using entity lookup and multi-hop traversal."""
        if not self.kg:
            return GraphRAGResult(strategy="local")

        # Extract entities from query
        entities = self._extract_query_entities(query)
        entity_context = []

        for entity in entities[:5]:
            try:
                triples = self.kg.query_entity(entity)
                if triples:
                    entity_context.append({
                        "entity": entity,
                        "facts": triples[:10],
                    })
            except Exception:
                continue

        # Try multi-hop between entity pairs
        if len(entities) >= 2:
            try:
                paths = self.kg.multi_hop(entities[0], entities[1], max_hops=3)
                if paths:
                    entity_context.append({
                        "entity": f"{entities[0]} → {entities[1]}",
                        "paths": paths[:3],
                    })
            except Exception:
                pass

        # Find matching communities for extra context
        matched_communities = []
        for entity in entities:
            cid = self._entity_to_community.get(entity.lower())
            if cid:
                community = next(
                    (c for c in self.communities if c.community_id == cid), None
                )
                if community and community not in matched_communities:
                    matched_communities.append(community)

        return GraphRAGResult(
            entity_context=entity_context,
            matched_communities=matched_communities[:3],
            strategy="local",
        )

    def _build_adjacency(self) -> Tuple[Dict[str, Set[str]], Dict[str, str]]:
        """Extract adjacency list from knowledge graph."""
        adjacency: Dict[str, Set[str]] = defaultdict(set)
        entity_names: Dict[str, str] = {}

        try:
            conn = self.kg._get_conn()

            # Get all entities
            rows = conn.execute("SELECT id, name FROM entities").fetchall()
            for r in rows:
                entity_names[r["id"]] = r["name"]

            # Get all triples
            rows = conn.execute(
                "SELECT subject_id, predicate, object_id FROM triples"
            ).fetchall()
            for r in rows:
                sid, oid = r["subject_id"], r["object_id"]
                adjacency[sid].add(oid)
                adjacency[oid].add(sid)

        except Exception as e:
            logger.warning(f"[GraphRAG] Failed to build adjacency: {e}")

        return adjacency, entity_names

    @staticmethod
    def _connected_components(
        adjacency: Dict[str, Set[str]],
        entity_names: Dict[str, str],
    ) -> List[Tuple[List[str], List[Tuple[str, str, str]]]]:
        """Find connected components in the graph."""
        visited: Set[str] = set()
        components = []

        for node in adjacency:
            if node in visited:
                continue

            # BFS for this component
            component_nodes: Set[str] = set()
            queue = [node]
            while queue:
                current = queue.pop(0)
                if current in visited:
                    continue
                visited.add(current)
                component_nodes.add(current)
                for neighbor in adjacency.get(current, set()):
                    if neighbor not in visited:
                        queue.append(neighbor)

            # Resolve to names
            names = [
                entity_names.get(n, n) for n in component_nodes
            ]

            # Collect relationships within this component
            relationships = []
            for n in component_nodes:
                for neighbor in adjacency.get(n, set()):
                    if neighbor in component_nodes:
                        subj = entity_names.get(n, n)
                        obj = entity_names.get(neighbor, neighbor)
                        relationships.append((subj, "related_to", obj))

            components.append((names, relationships))

        return components

    def _match_communities(self, query: str, top_k: int) -> List[Community]:
        """Match query to relevant communities via keyword overlap."""
        query_words = set(
            w.lower() for w in re.findall(r"\b\w{3,}\b", query)
        )

        scored = []
        for community in self.communities:
            community_words = set(
                w.lower() for w in community.keywords
            )
            # Also include entity names as words
            for entity in community.entities:
                community_words.update(
                    w.lower() for w in re.findall(r"\b\w{3,}\b", entity)
                )

            overlap = len(query_words & community_words)
            if overlap > 0:
                score = overlap / max(len(query_words), 1)
                scored.append((community, score))

        scored.sort(key=lambda x: x[1], reverse=True)
        return [c for c, _ in scored[:top_k]]

    def _generate_community_summary(self, community: Community) -> str:
        """Use LLM to generate a community summary."""
        try:
            entities_str = ", ".join(community.entities[:15])
            rels_str = "\n".join(
                f"  - {s} → {p} → {o}"
                for s, p, o in community.relationships[:20]
            )
            prompt = (
                f"The following entities and relationships form a thematic community "
                f"in a knowledge graph. Write a concise summary (2-4 sentences) "
                f"describing the main theme and key facts.\n\n"
                f"Entities: {entities_str}\n"
                f"Relationships:\n{rels_str}\n\n"
                f"Summary:"
            )
            response = self.llm_fn(prompt)
            if response and len(response.strip()) > 20:
                return response.strip()[:self.max_summary_len]
        except Exception as e:
            logger.debug(f"[GraphRAG] Summary generation failed: {e}")

        return self._heuristic_summary(community)

    @staticmethod
    def _heuristic_summary(community: Community) -> str:
        """Generate a simple summary without LLM."""
        entities = ", ".join(community.entities[:8])
        n_rels = len(community.relationships)
        return (
            f"This community contains {community.size} entities including "
            f"{entities}. There are {n_rels} known relationships among them."
        )

    @staticmethod
    def _extract_keywords(
        entities: List[str],
        relationships: List[Tuple[str, str, str]],
    ) -> List[str]:
        """Extract keywords from community entities and relationships."""
        words = set()
        for entity in entities:
            words.update(
                w.lower() for w in re.findall(r"\b\w{3,}\b", entity)
            )
        for s, p, o in relationships:
            words.update(
                w.lower() for w in re.findall(r"\b\w{3,}\b", f"{s} {p} {o}")
            )
        return sorted(words)

    @staticmethod
    def _extract_query_entities(query: str) -> List[str]:
        """Extract potential entity names from query."""
        # Capitalized multi-word phrases
        entities = re.findall(r"\b[A-Z][a-z]+(?:\s+[A-Z][a-z]+)*\b", query)
        # Quoted strings
        entities += re.findall(r'"([^"]+)"', query)
        entities += re.findall(r"'([^']+)'", query)

        if not entities:
            # Fall back to significant words
            entities = [
                w for w in re.findall(r"\b\w{4,}\b", query)
                if w.lower() not in {
                    "what", "where", "when", "which", "about",
                    "does", "have", "that", "this", "with",
                    "from", "they", "been", "were", "their",
                    "between", "relationship", "connection",
                }
            ]

        return entities[:5]

    @staticmethod
    def _detect_strategy(query: str) -> str:
        """Detect whether a query is global or local."""
        global_markers = [
            "main themes", "overview", "summarize", "key topics",
            "what are the", "list all", "major findings",
            "general", "overall", "themes", "categories",
        ]
        q_lower = query.lower()
        for marker in global_markers:
            if marker in q_lower:
                return "global"
        return "local"

File: rust_output/src/graph_rag.rs
/// Core/graph_rag::py — Graph RAG with Community Detection & Summarization.
/// 
/// Industry best practice: Beyond basic entity-triple storage, Graph RAG
/// builds community-level summaries enabling answers to global/thematic
/// queries that no single chunk can address.
/// 
/// Pipeline:
/// 1. Extract entities and relationships from chunks
/// 2. Build graph communities (connected components / Leiden)
/// 3. Generate summaries per community
/// 4. At query time: match communities → use summaries as context
/// 5. For local queries: traditional entity lookup + multi-hop
/// 
/// This is inspired by Microsoft's "GraphRAG" (2024) paper which showed
/// that community summaries dramatically improve answers to global
/// sensemaking questions ("What are the main themes in this dataset?").
/// 
/// References:
/// - Edge et al. "From Local to Global: A Graph RAG Approach" (Microsoft, 2024)
/// - Community detection via connected components (lightweight alternative to Leiden)

use anyhow::{Result, Context};
use serde::{Serialize, Deserialize};
use std::collections::HashMap;
use std::collections::HashSet;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// A community of related entities from the knowledge graph.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Community {
    pub community_id: String,
    pub entities: Vec<String>,
    pub relationships: Vec<(String, String, String)>,
    pub summary: String,
    pub keywords: Vec<String>,
    pub size: i64,
    pub level: i64,
}

/// Result from Graph RAG query.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphRAGResult {
    pub answer: String,
    pub matched_communities: Vec<Community>,
    pub entity_context: Vec<HashMap<String, Box<dyn std::any::Any>>>,
    pub strategy: String,
    pub community_summaries_used: i64,
}

/// Graph-based RAG with community detection and hierarchical summarization.
/// 
/// Extends the basic KnowledgeGraph with:
/// - Automatic community detection (connected components)
/// - LLM-generated community summaries
/// - Global query answering via community map-reduce
/// - Local query answering via entity + multi-hop traversal
/// 
/// Usage:
/// graph_rag = GraphRAG(knowledge_graph=kg, llm_fn=my_generate)
/// graph_rag::build_communities()  # After indexing documents
/// 
/// # Global query (themes, overview)
/// result = graph_rag::query("What are the main themes?", strategy="global")
/// 
/// # Local query (specific entity)
/// result = graph_rag::query("How does X relate to Y?", strategy="local")
#[derive(Debug, Clone)]
pub struct GraphRAG {
    pub kg: String,
    pub llm_fn: String,
    pub min_community_size: String,
    pub max_summary_len: String,
    pub communities: Vec<Community>,
    pub _entity_to_community: HashMap<String, String>,
}

impl GraphRAG {
    /// Args:
    /// knowledge_graph: KnowledgeGraph instance for entity/triple data
    /// llm_fn: function(prompt) -> str for summarization
    /// min_community_size: minimum entities to form a community
    /// max_community_summary_len: max chars per community summary
    pub fn new(knowledge_graph: Box<dyn std::any::Any>, llm_fn: Option<Box<dyn Fn>>, min_community_size: i64, max_community_summary_len: i64) -> Self {
        Self {
            kg: knowledge_graph,
            llm_fn,
            min_community_size,
            max_summary_len: max_community_summary_len,
            communities: Vec::new(),
            _entity_to_community: HashMap::new(),
        }
    }
    /// Detect communities via connected components and generate summaries.
    /// 
    /// Steps:
    /// 1. Extract adjacency from knowledge graph
    /// 2. Find connected components (lightweight community detection)
    /// 3. Filter by min size
    /// 4. Generate summaries per community
    pub fn build_communities(&mut self) -> Vec<Community> {
        // Detect communities via connected components and generate summaries.
        // 
        // Steps:
        // 1. Extract adjacency from knowledge graph
        // 2. Find connected components (lightweight community detection)
        // 3. Filter by min size
        // 4. Generate summaries per community
        if !self.kg {
            logger.warning("[GraphRAG] No knowledge graph provided".to_string());
            vec![]
        }
        let (mut adjacency, mut entity_names) = self._build_adjacency();
        if !adjacency {
            logger.info("[GraphRAG] Empty graph, no communities to build".to_string());
            vec![]
        }
        let mut components = self._connected_components(adjacency, entity_names);
        self.communities = vec![];
        self._entity_to_community = HashMap::new();
        for (i, (entities, relationships)) in components.iter().enumerate().iter() {
            if entities.len() < self.min_community_size {
                continue;
            }
            let mut cid = hashlib::sha256({ let mut v = entities.clone(); v.sort(); v }.join(&"|".to_string()).as_bytes().to_vec()).hexdigest()[..12];
            let mut community = Community(/* community_id= */ cid, /* entities= */ { let mut v = entities.clone(); v.sort(); v }, /* relationships= */ relationships, /* size= */ entities.len(), /* keywords= */ self._extract_keywords(entities, relationships));
            if self.llm_fn {
                community.summary = self._generate_community_summary(community);
            } else {
                community.summary = self._heuristic_summary(community);
            }
            self.communities.push(community);
            for entity in entities.iter() {
                self._entity_to_community[entity.to_lowercase()] = cid;
            }
        }
        logger.info(format!("[GraphRAG] Built {} communities from {} entities", self.communities.len(), self.communities.iter().map(|c| c.size).collect::<Vec<_>>().iter().sum::<i64>()));
        self.communities
    }
    /// Answer a query using the knowledge graph and community summaries.
    /// 
    /// Args:
    /// query: User query
    /// strategy: "local" (entity lookup), "global" (community summaries),
    /// or "auto" (detect best strategy)
    /// top_k_communities: max communities for global queries
    pub fn query(&mut self, query: String, strategy: String, top_k_communities: i64) -> GraphRAGResult {
        // Answer a query using the knowledge graph and community summaries.
        // 
        // Args:
        // query: User query
        // strategy: "local" (entity lookup), "global" (community summaries),
        // or "auto" (detect best strategy)
        // top_k_communities: max communities for global queries
        if strategy == "auto".to_string() {
            let mut strategy = self._detect_strategy(query);
        }
        if strategy == "global".to_string() {
            self._global_query(query, top_k_communities)
        } else {
            self._local_query(query)
        }
    }
    /// Answer using community summaries (map-reduce pattern).
    /// 
    /// 1. Match query to relevant communities
    /// 2. Collect community summaries
    /// 3. Use summaries as context for generation
    pub fn _global_query(&mut self, query: String, top_k: i64) -> Result<GraphRAGResult> {
        // Answer using community summaries (map-reduce pattern).
        // 
        // 1. Match query to relevant communities
        // 2. Collect community summaries
        // 3. Use summaries as context for generation
        let mut matched = self._match_communities(query, top_k);
        if !matched {
            GraphRAGResult(/* strategy= */ "global".to_string())
        }
        let mut summaries = vec![];
        for community in matched.iter() {
            if community.summary {
                summaries.push(format!("[Community: {}]\n{}", community.entities[..5].join(&", ".to_string()), community.summary));
            }
        }
        let mut context = summaries.join(&"\n\n".to_string());
        let mut answer = "".to_string();
        if (self.llm_fn && context) {
            // try:
            {
                let mut prompt = format!("Using the following knowledge graph community summaries, answer the question comprehensively.\n\nCommunity Summaries:\n{}\n\nQuestion: {}\n\nAnswer:", context, query);
                let mut answer = self.llm_fn(prompt);
            }
            // except Exception as e:
        }
        Ok(GraphRAGResult(/* answer= */ (answer || "".to_string()), /* matched_communities= */ matched, /* strategy= */ "global".to_string(), /* community_summaries_used= */ matched.len()))
    }
    /// Answer using entity lookup and multi-hop traversal.
    pub fn _local_query(&mut self, query: String) -> Result<GraphRAGResult> {
        // Answer using entity lookup and multi-hop traversal.
        if !self.kg {
            GraphRAGResult(/* strategy= */ "local".to_string())
        }
        let mut entities = self._extract_query_entities(query);
        let mut entity_context = vec![];
        for entity in entities[..5].iter() {
            // try:
            {
                let mut triples = self.kg.query_entity(entity);
                if triples {
                    entity_context.push(HashMap::from([("entity".to_string(), entity), ("facts".to_string(), triples[..10])]));
                }
            }
            // except Exception as _e:
        }
        if entities.len() >= 2 {
            // try:
            {
                let mut paths = self.kg.multi_hop(entities[0], entities[1], /* max_hops= */ 3);
                if paths {
                    entity_context.push(HashMap::from([("entity".to_string(), format!("{} → {}", entities[0], entities[1])), ("paths".to_string(), paths[..3])]));
                }
            }
            // except Exception as _e:
        }
        let mut matched_communities = vec![];
        for entity in entities.iter() {
            let mut cid = self._entity_to_community.get(&entity.to_lowercase()).cloned();
            if cid {
                let mut community = next(self.communities.iter().filter(|c| c.community_id == cid).map(|c| c).collect::<Vec<_>>(), None);
                if (community && !matched_communities.contains(&community)) {
                    matched_communities.push(community);
                }
            }
        }
        Ok(GraphRAGResult(/* entity_context= */ entity_context, /* matched_communities= */ matched_communities[..3], /* strategy= */ "local".to_string()))
    }
    /// Extract adjacency list from knowledge graph.
    pub fn _build_adjacency(&mut self) -> Result<(HashMap<String, HashSet<String>>, HashMap<String, String>)> {
        // Extract adjacency list from knowledge graph.
        let mut adjacency = defaultdict(set);
        let mut entity_names = HashMap::new();
        // try:
        {
            let mut conn = self.kg._get_conn();
            let mut rows = conn.execute("SELECT id, name FROM entities".to_string()).fetchall();
            for r in rows.iter() {
                entity_names[r["id".to_string()]] = r["name".to_string()];
            }
            let mut rows = conn.execute("SELECT subject_id, predicate, object_id FROM triples".to_string()).fetchall();
            for r in rows.iter() {
                let (mut sid, mut oid) = (r["subject_id".to_string()], r["object_id".to_string()]);
                adjacency[&sid].insert(oid);
                adjacency[&oid].insert(sid);
            }
        }
        // except Exception as e:
        Ok((adjacency, entity_names))
    }
    /// Find connected components in the graph.
    pub fn _connected_components(adjacency: HashMap<String, HashSet<String>>, entity_names: HashMap<String, String>) -> Vec<(Vec<String>, Vec<(String, String, String)>)> {
        // Find connected components in the graph.
        let mut visited = HashSet::new();
        let mut components = vec![];
        for node in adjacency.iter() {
            if visited.contains(&node) {
                continue;
            }
            let mut component_nodes = HashSet::new();
            let mut queue = vec![node];
            while queue {
                let mut current = queue.remove(&0);
                if visited.contains(&current) {
                    continue;
                }
                visited.insert(current);
                component_nodes.insert(current);
                for neighbor in adjacency.get(&current).cloned().unwrap_or(HashSet::new()).iter() {
                    if !visited.contains(&neighbor) {
                        queue.push(neighbor);
                    }
                }
            }
            let mut names = component_nodes.iter().map(|n| entity_names.get(&n).cloned().unwrap_or(n)).collect::<Vec<_>>();
            let mut relationships = vec![];
            for n in component_nodes.iter() {
                for neighbor in adjacency.get(&n).cloned().unwrap_or(HashSet::new()).iter() {
                    if component_nodes.contains(&neighbor) {
                        let mut subj = entity_names.get(&n).cloned().unwrap_or(n);
                        let mut obj = entity_names.get(&neighbor).cloned().unwrap_or(neighbor);
                        relationships.push((subj, "related_to".to_string(), obj));
                    }
                }
            }
            components.push((names, relationships));
        }
        components
    }
    /// Match query to relevant communities via keyword overlap.
    pub fn _match_communities(&mut self, query: String, top_k: i64) -> Vec<Community> {
        // Match query to relevant communities via keyword overlap.
        let mut query_words = re::findall("\\b\\w{3,}\\b".to_string(), query).iter().map(|w| w.to_lowercase()).collect::<Vec<_>>().into_iter().collect::<HashSet<_>>();
        let mut scored = vec![];
        for community in self.communities.iter() {
            let mut community_words = community.keywords.iter().map(|w| w.to_lowercase()).collect::<Vec<_>>().into_iter().collect::<HashSet<_>>();
            for entity in community.entities.iter() {
                community_words.extend(re::findall("\\b\\w{3,}\\b".to_string(), entity).iter().map(|w| w.to_lowercase()).collect::<Vec<_>>());
            }
            let mut overlap = (query_words & community_words).len();
            if overlap > 0 {
                let mut score = (overlap / query_words.len().max(1));
                scored.push((community, score));
            }
        }
        scored.sort(/* key= */ |x| x[1], /* reverse= */ true);
        scored[..top_k].iter().map(|(c, _)| c).collect::<Vec<_>>()
    }
    /// Use LLM to generate a community summary.
    pub fn _generate_community_summary(&mut self, community: Community) -> Result<String> {
        // Use LLM to generate a community summary.
        // try:
        {
            let mut entities_str = community.entities[..15].join(&", ".to_string());
            let mut rels_str = community.relationships[..20].iter().map(|(s, p, o)| format!("  - {} → {} → {}", s, p, o)).collect::<Vec<_>>().join(&"\n".to_string());
            let mut prompt = format!("The following entities and relationships form a thematic community in a knowledge graph. Write a concise summary (2-4 sentences) describing the main theme and key facts.\n\nEntities: {}\nRelationships:\n{}\n\nSummary:", entities_str, rels_str);
            let mut response = self.llm_fn(prompt);
            if (response && response.trim().to_string().len() > 20) {
                response.trim().to_string()[..self.max_summary_len]
            }
        }
        // except Exception as e:
        Ok(self._heuristic_summary(community))
    }
    /// Generate a simple summary without LLM.
    pub fn _heuristic_summary(community: Community) -> String {
        // Generate a simple summary without LLM.
        let mut entities = community.entities[..8].join(&", ".to_string());
        let mut n_rels = community.relationships.len();
        format!("This community contains {} entities including {}. There are {} known relationships among them.", community.size, entities, n_rels)
    }
    /// Extract keywords from community entities and relationships.
    pub fn _extract_keywords(entities: Vec<String>, relationships: Vec<(String, String, String)>) -> Vec<String> {
        // Extract keywords from community entities and relationships.
        let mut words = HashSet::new();
        for entity in entities.iter() {
            words.extend(re::findall("\\b\\w{3,}\\b".to_string(), entity).iter().map(|w| w.to_lowercase()).collect::<Vec<_>>());
        }
        for (s, p, o) in relationships.iter() {
            words.extend(re::findall("\\b\\w{3,}\\b".to_string(), format!("{} {} {}", s, p, o)).iter().map(|w| w.to_lowercase()).collect::<Vec<_>>());
        }
        { let mut v = words.clone(); v.sort(); v }
    }
    /// Extract potential entity names from query.
    pub fn _extract_query_entities(query: String) -> Vec<String> {
        // Extract potential entity names from query.
        let mut entities = re::findall("\\b[A-Z][a-z]+(?:\\s+[A-Z][a-z]+)*\\b".to_string(), query);
        entities += re::findall("\"([^\"]+)\"".to_string(), query);
        entities += re::findall("'([^']+)'".to_string(), query);
        if !entities {
            let mut entities = re::findall("\\b\\w{4,}\\b".to_string(), query).iter().filter(|w| !HashSet::from(["what".to_string(), "where".to_string(), "when".to_string(), "which".to_string(), "about".to_string(), "does".to_string(), "have".to_string(), "that".to_string(), "this".to_string(), "with".to_string(), "from".to_string(), "they".to_string(), "been".to_string(), "were".to_string(), "their".to_string(), "between".to_string(), "relationship".to_string(), "connection".to_string()]).contains(&w.to_lowercase())).map(|w| w).collect::<Vec<_>>();
        }
        entities[..5]
    }
    /// Detect whether a query is global or local.
    pub fn _detect_strategy(query: String) -> String {
        // Detect whether a query is global or local.
        let mut global_markers = vec!["main themes".to_string(), "overview".to_string(), "summarize".to_string(), "key topics".to_string(), "what are the".to_string(), "list all".to_string(), "major findings".to_string(), "general".to_string(), "overall".to_string(), "themes".to_string(), "categories".to_string()];
        let mut q_lower = query.to_lowercase();
        for marker in global_markers.iter() {
            if q_lower.contains(&marker) {
                "global".to_string()
            }
        }
        "local".to_string()
    }
}
